lib/backend/k8s/resources/kubeendpointslice.go

```diff
@@ -91,29 +94,44 @@ func (c *endpointSliceClient) Get(ctx context.Context, key model.Key, revision s
 
 func (c *endpointSliceClient) List(ctx context.Context, list model.ListInterface, revision string) (*model.KVPairList, error) {
 	log.Debug("Received List request on Kubernetes EndpointSlice type")
-	// List all of the k8s EndpointSlice objects.
-	endpointSlices := discovery.EndpointSliceList{}
-	req := c.clientSet.DiscoveryV1beta1().RESTClient().
-		Get().
-		Resource("endpointslices")
-	err := req.Do(ctx).Into(&endpointSlices)
-	if err != nil {
-		log.WithError(err).Info("Unable to list K8s EndpointSlice resources")
-		return nil, K8sErrorToCalico(err, list)
+	listFunc := func(ctx context.Context, opts metav1.ListOptions) (runtime.Object, error) {
+		return c.clientSet.DiscoveryV1beta1().EndpointSlices("").List(ctx, opts)
 	}
 
+	// For each object returned from the API server, add it to a KVPairList.
 	kvps := model.KVPairList{KVPairs: []*model.KVPair{}}
-	for _, es := range endpointSlices.Items {
-		kvp, err := c.EndpointSliceToKVP(&es)
+	kvps.Revision = revision
+	forEach := func(obj runtime.Object) error {
+		es := obj.(*discovery.EndpointSlice)
+		kvp, err := c.EndpointSliceToKVP(es)
 		if err != nil {
 			log.WithError(err).Info("Failed to convert K8s EndpointSlice")
-			return nil, err
+			return err
 		}
 		kvps.KVPairs = append(kvps.KVPairs, kvp)
+		return nil
+	}
+
+	lp := pager.New(listFunc)
+	opts := metav1.ListOptions{ResourceVersion: revision}
+	if revision != "" {
+		opts.ResourceVersionMatch = metav1.ResourceVersionMatchNotOlderThan
+	}
+	result, _, err := lp.List(ctx, opts)
+	if err != nil {
+		return nil, K8sErrorToCalico(err, list)
+	}
+	err = meta.EachListItem(result, forEach)
+	if err != nil {
+		return nil, K8sErrorToCalico(err, list)
 	}
 
-	// Add in the Revision information.
-	kvps.Revision = endpointSlices.ResourceVersion
+	// Extract the list revision information.
+	m, err := meta.ListAccessor(result)
+	if err != nil {
+		return nil, err
+	}
+	kvps.Revision = m.GetResourceVersion()
```

Comment:
This section of code starting from where a new pager is created down to here is almost the same in a few files - do you think it's worth extracting this into a helper function?